components/autocomplete/README.md

```diff
@@ -0,0 +1,121 @@
+
+### The Completer Interface
+
+#### name
+
+The name of the completer. Useful for identifying a specific completer to be overridden via extensibility hooks.
+
+- Type: `String`
+- Required: Yes
+
+#### options
+
+The raw options for completion. May be an array, a function that returns an array, or a function that returns a promise for an array.
+
+Options may be of any type or shape. The completer declares how those options are rendered and what their completions should be when selected.
+
+- Type: `Array|Function`
+- Required: Yes
+
+#### triggerPrefix
+
+The string prefix that should trigger the completer. For example, Gutenberg's block completer is triggered when the '/' character is entered.
+
+- Type: `String`
+- Required: Yes
+
+#### getOptionLabel
+
+A function that returns the label for a given option. A label may be a string or a mixed array of strings, elements, and components.
+
+- Type: `Function`
+- Required: Yes
+
+#### getOptionKeywords
+
+A function that returns the keywords for the specified option.
+
+- Type: `Function`
+- Required: Yes
+
+#### getOptionCompletion
+
+A function that takes an option and responds with how the option should be completed. By default, the result is a value to be inserted in the text. However, a completer may explicitly declare how a completion should be treated by returning an object with `action` and `value` properties. The `action` declares what should be done with the `value`.
+
+There are currently two supported actions:
+
+* "insert-at-caret" - Insert the `value` into the text (the default completion action).
+* "replace" - Replace the current block with the block specified in the `value` property.
+
+#### allowNode
+
+A function that takes a text node and returns a boolean indicating whether the completer should be considered for that node.
+
+- Type: `Function`
+- Required: No
+
+#### allowContext
+
+A function that takes a Range before and a Range after the autocomplete trigger and query text and returns a boolean indicating whether the completer should be considered for that context.
+
+- Type: `Function`
+- Required: No
+
+#### className
+
+A class name to apply to the autocompletion popup menu.
+
+- Type: `String`
+- Required: No
+
+### Examples
+
+The following is a contrived completer for fresh fruit.
+
+```jsx
+const fruitCompleter = {
+	name: 'fruit',
+	// The prefix that triggers this completer
+	triggerPrefix: '~',
+	// The option data
+	options: [
+		{ visual: '🍎', name: 'Apple' },
+		{ visual: '🍊', name: 'Orange' ,
+		{ visual: '🍇', name: 'Grapes' },
+	],
+	// Returns a label for an option like "🍊 Orange"
+	getOptionLabel: option => [
+		<span class="icon">{ option.visual }</span>,
+		option.name
+	],
+	// Declares that options should be matched by their name
+	getOptionKeywords: option => [ option.name ],
+	// Declares completions should be inserted as abbreviations
+	getOptionCompletion: option => (
+		<abbr title={ option.name }>{ option.visual }</abbr>
+	),
+};
+```
```

Comment:
Can we also add a line where it is wired into an actual component?

Reply:
I created a simple example wiring to `Autocomplete` wrapping an `input` element, but it didn't work. It seems like `Autocomplete` is assuming a contenteditable target, since it inserts HTML rather than simple text. I'll play with a simple contenteditable example but don't have anything yet.